wallet: include a checksum of encrypted private keys

achow101 opened the pull request:
Adds a checksum to the encrypted key record in the wallet database so that encrypted keys can be checked for corruption on wallet loading, in the same way that unencrypted keys are. This allows for us to skip the full decryption of keys upon the first unlocking of the wallet in that session as any key corruption will have already been detected. The checksum is just the double SHA256 of the encrypted key and it is appended to the record after the encrypted key itself.

This is backwards compatible as old wallets will be able to read the encrypted key and ignore that there is more data in the stream. Additionally, old wallets will be upgraded upon their first unlocking (so that key decryption is checked before we commit to a checksum of the encrypted key) and a wallet flag set indicating that. The presence of the wallet flag lets us skip the full decryption as if `fDecryptionThoroughlyChecked` were true.

This does mean that the first time an old wallet is unlocked in a new version will take much longer, but subsequent unlocks will be instantaneous. Furthermore, corruption will be detected upon loading rather than on trying to send so wallet corruption will be detected sooner.

Fixes #12423

achow101 committed:
Read and write a checksum for encrypted keys

practicalswift commented:
I'm afraid the current version of this PR will introduce a bug in `WalletBatch::WriteCryptedKey`:

https://github.com/bitcoin/bitcoin/blob/2b79a270e7442309ab167b161384060d93808695/src/wallet/walletdb.cpp#L118-L128

Please note the `return false;` which makes the re-try unreachable :)

achow101 commented:
> I'm afraid the current version of this PR will introduce a bug in `WalletBatch::WriteCryptedKey`:
> 
> Please note the first `return false;` which makes the re-try logic unreachable :)

Whoops, fixed.

laanwj commented:
Concept ACK

luke-jr commented:
> a wallet flag set indicating that. The presence of the wallet flag lets us skip the full decryption as if fDecryptionThoroughlyChecked were true.

What happens if I open with new version, then switch back to an old version, generate some new keys, then switch back to the new?

Is a wallet flag really needed at all? Shouldn't verifying checksums also reveal if there are checksums missing?

achow101 commented:
This is currently getting messed up by the ScriptPubKeyMan changes. Will fix, rebase, and address comments after ScriptPubKeyMan is fully introduced.

achow101 committed:
Set fDecryptionThoroughlyChecked based on whether crypted key checksums are valid

Change fDecryptionThoroughlyChecked to default to true so that it can
latch to false when an invalid checksum is seen. Checksums may be invalid
if the wallet does not have checksums or if the wallet became corrupted.

It is safe to default fDecryptionThoroughlyChecked to true because any
existing wallet without a checksum will set it to false. Any new or
blank wallet where encrypted keys are added will then set this to true
when the first encrypted key is generated by virtue of CheckDecryptionKey
doing that during the initial Unlock prior to keys being added.

achow101 committed:
Upgrade or rewrite encrypted key checksums

If fDecryptionThoroughlyChecked is false, after a key has been checked,
write (or rewrite) its checksum. This serves to upgrade wallets
and correct those which have the checksum corrupted but not the key.

achow101 commented:
I've rebased onto master and included #17369 as a prerequisite. I've also reworked it according to @luke-jr's suggestions. Instead of a wallet flag, we reuse `fDecryptionThoroughlyChecked` to let as avoid the decryption check. This also makes it easier for us to write (or rewrite) the checksums if they don't exist (or are invalid).

laanwj commented:
Might want to rebase to get rid of the commits from #17369 and make it easier to review.

achow101 commented:
Rebased

jonatack reviewed (commented):
Code review ACK d67055e00dd90f504384e5c3f229fc95306d5aac 

Rebased on current master @ 5236b2e, built, ran tests, ran bitcoind.

Could use test coverage.

meshcollider reviewed (commented):
Code review ACK d67055e00dd90f504384e5c3f229fc95306d5aac

achow101 commented:
ping  @instagibbs

laanwj commented:
code review ACK d67055e00dd90f504384e5c3f229fc95306d5aac
